torch/distributed/fsdp/flat_param.py

```diff
@@ -354,67 +366,98 @@ def __init__(
         use_orig_params: bool,
     ):
         super().__init__()
-        use_unsafe_setattr = os.environ.get(_FSDP_USE_UNSAFE_SETATTR, "") == "1"
-        self._setattr_tensor: Callable[[nn.Module, str, Tensor], None]
-        self._setattr_param: Callable[[nn.Module, str, nn.Parameter], None]
-        if use_unsafe_setattr:
-            self._setattr_tensor = _unsafe_setattr_tensor
-            self._setattr_param = _unsafe_setattr_param
-        else:
-            self._setattr_tensor = _safe_setattr_tensor_or_param
-            self._setattr_param = _safe_setattr_tensor_or_param
+        params = list(params)
+        if len(params) == 0:
+            raise ValueError(
+                f"Cannot construct a {self.__class__.__name__} with an empty parameter list"
+            )
+        self._init_setattr_fns()
+        align_addresses = use_orig_params
+        self._init_get_unflat_views_fn(align_addresses)
+
         self.device = device
         self.process_group = process_group
         self.rank = process_group.rank()
         self.world_size = process_group.size()
+
         self._sharding_strategy = sharding_strategy
         self._offload_params = offload_params
         self._use_orig_params = use_orig_params
         self._keep_low_precision_grads = keep_low_precision_grads
         self._training_state = HandleTrainingState.IDLE
         self._debug_level = dist.get_debug_level()
         self._fully_sharded_module = fully_sharded_module
-        self._init_flat_param(params, fully_sharded_module, use_orig_params)
-        self._orig_param_dtype = self.flat_param.dtype
-        self._use_unsharded_views(as_params=False)
+
+        # Optimistically assume a valid input `params` and set dtype attributes
+        # before `_init_flat_param()`, which performs the actual validation
+        self._orig_param_dtype = params[0].dtype
         self._init_param_reduce_dtypes(mp_param_dtype, mp_reduce_dtype)
+        assert self._fwd_bwd_param_dtype is not None  # mypy
+        self._aligned_numel = (
+            _get_aligned_numel(unsharded_dtype=self._fwd_bwd_param_dtype)
```

Comment:
why do we have to distinguish btwn unsharded_dtype vs sharded_dtype here, wouldn't they be the same?